### PR TITLE
Refactor Layer Structure

### PR DESCRIPTION
**Pull request by sharvenp**
## Changes in this PR
- Implemented proposed layer refactor in this [discussion](https://github.com/ramp4-pcar4/ramp4-pcar4/discussions/658) post

[Demo](http://ramp4-app.azureedge.net/demo/users/sharvenp/layer-refactor/host/index.html)
[Demo index-e2e](http://ramp4-app.azureedge.net/demo/users/sharvenp/layer-refactor/host/index-e2e.html)

[HackMD Doc](https://hackmd.io/@sharvenp/HyFbeXmrt/edit) contains some info on design choices

---
This change is [Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2)

### CODE REVIEW

**Review by james-rae — CHANGES_REQUESTED** [review]
> Reviewed 24 of 38 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2)* status: 24 of 38 files reviewed, 5 unresolved discussions (waiting on @james-rae and @sharvenp)

___
*[packages/ramp-core/src/geo/layer/attrib-layer.ts, line 47 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MonjB_q8Ipg8kJGeZ6M:-MonjB_q8Ipg8kJGeZ6N:b-ghagrr) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/attrib-layer.ts#L47)):*
> ```XML
>     // @ts-ignore
>     geomType: GeometryType;
>     esrifields: Array<EsriField>;
> ```

not entirely sure how this is being used, but since public, possible to use `FieldDefinition` instead?
Ah i see we have `.fields` that is that. So maybe just change the case to `esriFields` and this is in line with our other naughty `.esriXYZ` properties.
___
*[packages/ramp-core/src/geo/layer/attrib-layer.ts, line 49 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monio-V446lGqpReQLm:-Monio-V446lGqpReQLn:byjocb4) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/attrib-layer.ts#L49)):*
> ```XML
>     esrifields: Array<EsriField>;
>     fieldList: string; // list of field names, useful for numerous esri api calls
>     extent: EsriExtent | undefined;
> ```

since this is public, should probably be `Extent` from `/geo/api`
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 55 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MontY7n3oR4EMPUV62u:-MontY7n3oR4EMPUV62v:bve4caf) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L55)):*
> ```XML
> 
>     // TODO[Layer-Refactor]: making public so parent layer can push watches
>     watches: Array<__esri.WatchHandle>;
> ```

ok with public. wondering do we call it `.esriWatches` for consistency? Seems a bit wordy, but also something that won't be referenced often. I'm on the fence.
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 134 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monu50u7JigNkxR53CV:-Monu50u7JigNkxR53CW:bejfxct) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L134)):*
> ```XML
> 
>         if (this.isSublayer) {
>             // early back out, we don't want the below code to run for sublayers
> ```

would a console warn be appropriate here?
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 338 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monv8J1CLUBv1W0F6R1:-Monv8J1CLUBv1W0F6R2:bt2dnkg) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L338)):*
> ```XML
> 
>             this.sublayers.forEach(
>                 sublayer => (sublayer as any)?.onLoad() // This will just trigger the above statements
> ```

could `any` be `LayerInstance`?  Alternatley, since we are clearly expecting this function, maybe it should be on `LayerInstance` definition (default declaration can be a do-nothing function).
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 480 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MonxXXa2rRGC622nAJl:-MonxXXb0kRWha4WUnbc:b-wzcdd) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/64e380ddc5b17204a309e977caa8a66a2a57dacf/packages/ramp-core/src/geo/layer/common-layer.ts#L480)):*
> ```XML
>      * @returns {BaseFC} the matching feature class object, or undefined if the root was requested
>      */
>     protected getFC(
> ```

RIP you glorious function
___
*[packages/ramp-core/src/geo/layer/file-layer.ts, line 63 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monz3aPA1X5xYzg-oyQ:-Monz3aPA1X5xYzg-oyR:b-qpk59n) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/file-layer.ts#L63)):*
> ```XML
> 
> export class FileLayer extends AttribLayer {
>     // TODO: declare was added as part of the layer-state refactor - do we need this?
> ```

probably. I have it peppered in other spots. it is a nice keyword, very respectable.

**Review by james-rae — CHANGES_REQUESTED** [review]
FC files can be deleted once ur happy.

The code that uses layers looks much cleaner now, which hopefully means the idea was ok.

> Reviewed 14 of 38 files at r1.
> *[Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-:-MooFmJX6uQyu2X5GsKx:b-k2rwu9)* status: all files reviewed, 11 unresolved discussions (waiting on @sharvenp)

___
*[packages/ramp-core/src/geo/layer/layer.ts, line 375 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooElozEc4BSCN-LINa:-MooElozEc4BSCN-LINb:bjjo8fc) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/layer.ts#L375)):*
> ```XML
>      */
>     id: string;
>     uid: string;
> ```

guess i'm a bit confused why the implementation moved to this class from `CommonLayer`. `LayerInstance` has traditionally been an "class acting like an interface", which allowed us the flexibility of making a new layer type totally disconnected from all the implementation rooted in `CommonLayer`.   Open to discuss, but at a glance I'd like to keep the stubs with errors/not implemented warnings and have `CommonLayer` do the implementation.
___
*[packages/ramp-core/src/geo/layer/esriMapImage/index.ts, line 286 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooI0JP0Am0YiB3m8l9:-MooI0JP0Am0YiB3m8lA:b-y1a2o1) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/esriMapImage/index.ts#L286)):*
> ```XML
>                     );
>                 }
>                 if (this._sublayers[sid].isRemoved) {
> ```

this array lookup is repeated many times. can we just assign the pointer to a local var? (yes orig code did it too, bad james!)
___
*[packages/ramp-core/src/geo/layer/esriMapImage/index.ts, line 626 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooIpWj8rc8HZZn7d56:-MooIpWj8rc8HZZn7d57:bqremtm) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/esriMapImage/index.ts#L626)):*
> ```XML
>         }
> 
>         const activeFCs: Array<MapImageSublayer> = options.sublayerIds
> ```

rename var
___
*[packages/ramp-core/src/geo/map/ramp-map.ts, line 302 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooLGPZ3SSSQiy_GNSn:-MooLGPZ3SSSQiy_GNSo:b-5p1g9g) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/map/ramp-map.ts#L302)):*
> ```XML
> 
>     /**
>      * Removes a sublayer from the legend
> ```

change "legend" to "map". This module should be unaware the legend exists
___
*[packages/ramp-core/src/geo/map/ramp-map.ts, line 322 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooLSiv6TAqvEwliwsP:-MooLSiv6TAqvEwliwsQ:b-hx1rzt) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/map/ramp-map.ts#L322)):*
> ```XML
>             layer = sublayer;
>         }
>         this.$iApi.$vApp.$store.dispatch(LegendStore.removeLayerEntry, uid);
> ```

this is assuming a legend exists, and now has the map manipulating the legend. is there a way we can break the hard-link via an event (the one below, or a dedicated `SUBLAYER_REMOVE` event?)
___
*[packages/ramp-core/src/store/modules/layer/layer-store.ts, line 41 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooMNQJ2cJco-Yy4g6V:-MooMNQJ2cJco-Yy4g6W:bq3xs84) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/store/modules/layer/layer-store.ts#L41)):*
> ```XML
>             while (queue.length > 0) {
>                 const layer = queue.shift();
>                 if (layer?.uid === uid) {
> ```

if you want to get fancy, you could abstract these twin BFS algo's and have this one comparison operator become a param. I don't care, just pointing out an opportunity for amazing fun.

**Comment by james-rae** [comment]
___
*[packages/ramp-core/src/geo/layer/layer.ts, line 375 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooElozEc4BSCN-LINa:-MooOWYP4YqAHcCJ1GSI:b-8z78uk) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/layer.ts#L375)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

guess i'm a bit confused why the implementation moved to this class from `CommonLayer`. `LayerInstance` has traditionally been an "class acting like an interface", which allowed us the flexibility of making a new layer type totally disconnected from all the implementation rooted in `CommonLayer`.   Open to discuss, but at a glance I'd like to keep the stubs with errors/not implemented warnings and have `CommonLayer` do the implementation.
</blockquote></details>

This comment looks weird in collapsed view, but I'm referring to pretty much all the code changes in this file. The underscore property storage, the utility functions, etc.

**Review by sharvenp — COMMENTED** [review]
Yup while updating some of the code in the Vue components, I can definitely see things getting simpler.

I'll remove all block comments and FC files before making a PR into the R4 main repo to avoid spamming this PR with massive changes 👍

> *[Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-:-Mp7WDIIDS8kOvrxj3nU:b-rwag3w)* status: 0 of 41 files reviewed, 11 unresolved discussions (waiting on @james-rae)

___
*[packages/ramp-core/public/starter-scripts/panel-party.js, line 45 at r3](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Mp7eb87CdlAv8hc5VpT:-Mp7eb87CdlAv8hc5VpU:b-ao45pd) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/8035fe63a807e910ccd215fb4b060963ee54dd08/packages/ramp-core/public/starter-scripts/panel-party.js#L45)):*
> ```JavaScript
>                     extentSetId: 'EXT_ESRI_World_AuxMerc_3857',
>                     lodSetId: 'LOD_ESRI_World_AuxMerc_3857'
>                     // thumbnailTileUrls: ['/tile/8/91/74', '/tile/8/91/75']
> ```

I rebased Elsa's PR from `master` into my branch, hence the changes in this file. 
___
*[packages/ramp-core/src/fixtures/basemap/item.vue, line 33 at r3](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Mp7ewB29GyUDCQtJUj9:-Mp7ewB29GyUDCQtJUjA:b2nreh6) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/8035fe63a807e910ccd215fb4b060963ee54dd08/packages/ramp-core/src/fixtures/basemap/item.vue#L33)):*
> ```Vue
>                         />
>                         
>                          ```

Same as above.
___
*[packages/ramp-core/src/geo/layer/attrib-layer.ts, line 47 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MonjB_q8Ipg8kJGeZ6M:-Mp7X1mA4uzifgQLsfre:bah2568) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/attrib-layer.ts#L47)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

not entirely sure how this is being used, but since public, possible to use `FieldDefinition` instead?
Ah i see we have `.fields` that is that. So maybe just change the case to `esriFields` and this is in line with our other naughty `.esriXYZ` properties.
</blockquote></details>

Donethanks
___
*[packages/ramp-core/src/geo/layer/attrib-layer.ts, line 49 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monio-V446lGqpReQLm:-Mp7YAtuC1YpdSo2x-dY:b-r11slu) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/attrib-layer.ts#L49)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

since this is public, should probably be `Extent` from `/geo/api`
</blockquote></details>

This [line](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/8035fe63a807e910ccd215fb4b060963ee54dd08/packages/ramp-core/src/geo/layer/file-layer.ts#L358) in `FileLayer` seems to be assigning an `EsriExtent` to this property and so changing it to `Extent` from `/geo/api` causes an error. Should I make this `private/protected`?
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 55 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MontY7n3oR4EMPUV62u:-Mp7XOZi2BzVKgrUvH0o:b-jcdx1r) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L55)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

ok with public. wondering do we call it `.esriWatches` for consistency? Seems a bit wordy, but also something that won't be referenced often. I'm on the fence.
</blockquote></details>

`.esriWatches` sounds good. Donethanks
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 134 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monu50u7JigNkxR53CV:-Mp7ZCGB4wQWKB9hyMkp:bvlxq2b) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L134)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

would a console warn be appropriate here?
</blockquote></details>

Yes adding a `console.warn` makes sense. Donethanks
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 338 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monv8J1CLUBv1W0F6R1:-Mp7aMU68beuoSlDSvUa:b-b9p8o7) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L338)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

could `any` be `LayerInstance`?  Alternatley, since we are clearly expecting this function, maybe it should be on `LayerInstance` definition (default declaration can be a do-nothing function).
</blockquote></details>

It now works with `CommonLayer` instead of `any`. I tried adding an empty `OnLoad` function to `LayerInstance` but I can only make it `protected` since `CommonLayer` uses a `protected` `OnLoad`. But if I make it `protected`, then I cannot make the call below. What would you suggest I do in this case? Should I make `OnLoad` in `CommonLayer` not `protected`?
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 480 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MonxXXa2rRGC622nAJl:-Mp7b0eo1xqk6dkOIPX0:bhyf06x) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/64e380ddc5b17204a309e977caa8a66a2a57dacf/packages/ramp-core/src/geo/layer/common-layer.ts#L480)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

RIP you glorious function
</blockquote></details>

Reincarnated as [`getSublayer`](https://github.com/sharvenp/ramp4-pcar4/blob/5fac9fa1eb807fd50294ca00a9c9882cdb17332e/packages/ramp-core/src/geo/layer/layer.ts#L920) :)
___
*[packages/ramp-core/src/geo/layer/file-layer.ts, line 63 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monz3aPA1X5xYzg-oyQ:-Mp7bJ535M6rpWFwVLJR:bah2568) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/file-layer.ts#L63)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

probably. I have it peppered in other spots. it is a nice keyword, very respectable.
</blockquote></details>

Donethanks
___
*[packages/ramp-core/src/geo/layer/layer.ts, line 375 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooElozEc4BSCN-LINa:-Mp7coxOAnAY1jgAY3bh:b-3qupin) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/layer.ts#L375)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

This comment looks weird in collapsed view, but I'm referring to pretty much all the code changes in this file. The underscore property storage, the utility functions, etc.
</blockquote></details>

I completely agree and I've moved most of the properties and methods back into `CommonLayer`. 
In regards to the error stubs in `CommonLayer`, the default implementation for the getters/setters are used by the child classes for property management. I think this is cleaner since they all use the same interface for reading and updating property values. If we want to go around this, then we'd access the property directly in the subclass instead of using the getters/setters (e.g. use `this._name = ...` instead of `this.name = ...`). 
I am okay with both. 
___
*[packages/ramp-core/src/geo/layer/esriMapImage/index.ts, line 286 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooI0JP0Am0YiB3m8l9:-Mp7bx6r4bcbMUE0keKY:bah2568) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/esriMapImage/index.ts#L286)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

this array lookup is repeated many times. can we just assign the pointer to a local var? (yes orig code did it too, bad james!)
</blockquote></details>

Donethanks
___
*[packages/ramp-core/src/geo/layer/esriMapImage/index.ts, line 626 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooIpWj8rc8HZZn7d56:-Mp7blADAjTttxohJ0Oc:bah2568) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/esriMapImage/index.ts#L626)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

rename var
</blockquote></details>

Donethanks
___
*[packages/ramp-core/src/geo/map/ramp-map.ts, line 302 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooLGPZ3SSSQiy_GNSn:-Mp7cJG76KjUHHysBi-m:bah2568) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/map/ramp-map.ts#L302)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

change "legend" to "map". This module should be unaware the legend exists
</blockquote></details>

Donethanks
___
*[packages/ramp-core/src/geo/map/ramp-map.ts, line 322 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooLSiv6TAqvEwliwsP:-Mp7cKsHEzDz82TUngLy:b-dx84vi) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/map/ramp-map.ts#L322)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

this is assuming a legend exists, and now has the map manipulating the legend. is there a way we can break the hard-link via an event (the one below, or a dedicated `SUBLAYER_REMOVE` event?)
</blockquote></details>

Removed this call because it's unnecessary and the handler for the below `LAYER_REMOVE` event will run this exact line. Donethanks
___
*[packages/ramp-core/src/store/modules/layer/layer-store.ts, line 41 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooMNQJ2cJco-Yy4g6V:-Mp7bzYP8CRZyIPTMm5o:b-qh5qtc) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/store/modules/layer/layer-store.ts#L41)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

if you want to get fancy, you could abstract these twin BFS algo's and have this one comparison operator become a param. I don't care, just pointing out an opportunity for amazing fun.
</blockquote></details>

My algorithms professor will be proud 👨‍🏫

**Review by james-rae — CHANGES_REQUESTED** [review]
Sounds good.  Silly but i'm super happy you reverted back to `layerIdx`, less letters for the lazy win!!!

> Reviewed 2 of 40 files at r2, 38 of 39 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-:-MpI6Nj0Cm0WluazKG-5:b-nb6miz)* status: all files reviewed, 4 unresolved discussions (waiting on @james-rae and @sharvenp)

___
*[packages/ramp-core/src/geo/api/geo-defs.ts, line 140 at r4](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MpI6ciU0j0TZvacAhPN:-MpI6ciU0j0TZvacAhPO:blbscaa) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/5fac9fa1eb807fd50294ca00a9c9882cdb17332e/packages/ramp-core/src/geo/api/geo-defs.ts#L140)):*
> ```XML
>     UNKNOWN = 'unknown',
> 
>     SUBLAYER = 'sublayer' // TODO[Layer-Refactor]: general purpose sublayer layer type - is this okay?
> ```

its ok (since we need something). My only mild concern is in the doc pages, as we do not want to imply people can make config nuggets with `sublayer` as a valid type. But thats an issue for Markdown for us to deal with later, me thinks.
___
*[packages/ramp-core/src/geo/layer/attrib-layer.ts, line 49 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monio-V446lGqpReQLm:-MpI8QH6-HyUvFYb4LSt:b9mvftx) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/attrib-layer.ts#L49)):*
<details><summary><i>Previously, sharvenp (Sharven Prasad Dhanasekar) wrote…</i></summary><blockquote>

This [line](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/8035fe63a807e910ccd215fb4b060963ee54dd08/packages/ramp-core/src/geo/layer/file-layer.ts#L358) in `FileLayer` seems to be assigning an `EsriExtent` to this property and so changing it to `Extent` from `/geo/api` causes an error. Should I make this `private/protected`?
</blockquote></details>

my preferred solution is to change that line to be something like (am just guessing the actual path, code will need correction)
```
this.extent = this.$iApi.geo.geometry.geomEsriToRamp(l.fullExtent, layerId + '_extent');
```

since this in internal, you could shortcut into the subfunction `_convEsriExtentToRamp()` to avoid casting complaining.

changing the type on `.extent` might cascade into other changes, so if it starts to looks crazy then PM me and we can reconsider
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 338 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monv8J1CLUBv1W0F6R1:-MpIA0Fh5h7hDbaJxSpL:b-79195) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L338)):*
<details><summary><i>Previously, sharvenp (Sharven Prasad Dhanasekar) wrote…</i></summary><blockquote>

It now works with `CommonLayer` instead of `any`. I tried adding an empty `OnLoad` function to `LayerInstance` but I can only make it `protected` since `CommonLayer` uses a `protected` `OnLoad`. But if I make it `protected`, then I cannot make the call below. What would you suggest I do in this case? Should I make `OnLoad` in `CommonLayer` not `protected`?
</blockquote></details>

hmmmm. dilly of a pickle. removing protected makes the types nicer. but then risks cool kids wanting to call it. 

Now that I look at the full code, it appears `onLoad()` only exists in commonlayer and it is `onLoadActions()` that gets customized. So maybe going back to `any` is fine. Not sure if worth putting in a check & error if `onLoad()` doesn't exists.... would appear that anything that inherits from common layer will have it.  Third party layers outside of the heirarchy can react to the esri load events in ways they please.

So would appear my anger at any `any` may not be worth it here. Your call.
___
*[packages/ramp-core/src/geo/layer/layer.ts, line 375 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooElozEc4BSCN-LINa:-MpIEU362p-wzVdb672v:b48mhye) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/layer.ts#L375)):*
<details><summary><i>Previously, sharvenp (Sharven Prasad Dhanasekar) wrote…</i></summary><blockquote>

I completely agree and I've moved most of the properties and methods back into `CommonLayer`. 
In regards to the error stubs in `CommonLayer`, the default implementation for the getters/setters are used by the child classes for property management. I think this is cleaner since they all use the same interface for reading and updating property values. If we want to go around this, then we'd access the property directly in the subclass instead of using the getters/setters (e.g. use `this._name = ...` instead of `this.name = ...`). 
I am okay with both. 
</blockquote></details>

^ if thats a question, I'm ok with whatever you like
___
*[packages/ramp-core/src/geo/layer/esriMapImage/map-image-sublayer.ts, line 10 at r4](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MpIG3jr4ZsV5vNlaA_q:-MpIG3jr4ZsV5vNlaA_r:b-mj5ee2) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/5fac9fa1eb807fd50294ca00a9c9882cdb17332e/packages/ramp-core/src/geo/layer/esriMapImage/map-image-sublayer.ts#L10)):*
> ```XML
> export class MapImageSublayer extends AttribLayer {
>     // @ts-ignore
>     // esriSubLayer: __esri.Sublayer; //moved to layer instance //TODO[Layer-Refactor]: is this okay?
> ```

i think so

**Review by sharvenp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2)* status: 36 of 41 files reviewed, 4 unresolved discussions (waiting on @james-rae)

___
*[packages/ramp-core/src/geo/layer/attrib-layer.ts, line 49 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monio-V446lGqpReQLm:-MpITEaY-RhyAFDnoDtZ:b-53h04f) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/attrib-layer.ts#L49)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

my preferred solution is to change that line to be something like (am just guessing the actual path, code will need correction)
```
this.extent = this.$iApi.geo.geometry.geomEsriToRamp(l.fullExtent, layerId + '_extent');
```

since this in internal, you could shortcut into the subfunction `_convEsriExtentToRamp()` to avoid casting complaining.

changing the type on `.extent` might cascade into other changes, so if it starts to looks crazy then PM me and we can reconsider
</blockquote></details>

The change works and there's no cascading to worry about. Donethanks
___
*[packages/ramp-core/src/geo/layer/common-layer.ts, line 338 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-Monv8J1CLUBv1W0F6R1:-MpITlwE2WqmKYozo6z4:bq5oz59) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/common-layer.ts#L338)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

hmmmm. dilly of a pickle. removing protected makes the types nicer. but then risks cool kids wanting to call it. 

Now that I look at the full code, it appears `onLoad()` only exists in commonlayer and it is `onLoadActions()` that gets customized. So maybe going back to `any` is fine. Not sure if worth putting in a check & error if `onLoad()` doesn't exists.... would appear that anything that inherits from common layer will have it.  Third party layers outside of the heirarchy can react to the esri load events in ways they please.

So would appear my anger at any `any` may not be worth it here. Your call.
</blockquote></details>

Reverted back to `any`. I also added the function check since `sublayer` could be a custom `LayerInstance` with no `OnLoad`. Not really sure if this case will ever occur but it's good to be safe. Donethanks
___
*[packages/ramp-core/src/geo/layer/layer.ts, line 375 at r1](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2#-MooElozEc4BSCN-LINa:-MpIU9DJ1h6NodmGfgvR:b-pmx8fc) ([raw file](https://github.com/james-rae/ramp4-pcar4/blob/7a1c7d64524749bffe0fb3e4f58e40f09ebbb51c/packages/ramp-core/src/geo/layer/layer.ts#L375)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

^ if thats a question, I'm ok with whatever you like
</blockquote></details>

I decided to keep it as it is where `CommonLayer` doesn't throw error stubs during common layer property management. This way child classes can manage their properties using the same interface.

**Review by james-rae — APPROVED** [review]
> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2)* status: all files reviewed, 1 unresolved discussion (waiting on @james-rae)

**Review by james-rae — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/james-rae/ramp4-pcar4/2)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @sharvenp)
